README.md
![ERC-721 NES Tests](https://github.com/samurisenft/erc721nes-contracts/actions/workflows/tests.yml/badge.svg)

# About this project

ERC-721 NES, or Non Escrow Staking, is a novel implementation of a staking model that does not require the owner of a token to lock it into an escrow contract. What that means is the token never moves during the staking process, unlike traditional protocols where you receive a coupon to then redeem later for your staked tokens. Instead, the token is flagged as non transferrable for the entire staking duration. In this way, the owner of a token has zero exposure to the risk of a staking protocol being compromised and their tokens stolen.

If a protocol is not using your token to create some sort of financial derivative, it has no business in having the permission of ownership over it.

Our hypothesis is that this distinction is an important one for blockchain gaming. As a token grows in value, the token owner is inversely incentivized for engagement: their token is locked in an escrow contract that could potentially lose their token in the case of faulty code.

NES also offers other ancillary advantages, like being able to receive purchase offers on tokens from various marketplaces or being able to easily prove ownership of a token for things like Collab.Land, Premint or Twitter.

# Implementation

ERC-721 NES provides an interface to staking protocols by decorating the prototypical ERC-721 contract with a few additional methods and maintaining one new piece of state. Now, instead of a custodial approach where the owner relinquishes all the power over their token, a signal based locking mechanism can be employed.

```
_stake(uint256 tokenId)
// locks the token

_unstake(uint256 tokenId)
// unlocks the token

isStaked(uint256 tokenId)
// returns a boolean indicating the staked status of the token

address stakingController
// address of the staking contract
```

# Getting Started

## Integrate
- Install ERC-721 NES as a dependency via NPM  
  ```
  npm install erc721nes
  ```

- Extend your ERC-721 contract using ERC721NES  
  ```
  import "erc721nes/ERC721NES.sol";
  
  contract MyNFT is ERC721NES {
    // your ERC-721 code here
  }
  ```

# Building and Installing Locally
1. Fork the repo 
2. git clone [your fork url]
3. Star the project ;-)
4. Run these command locally:
```
npm install
npx hardhat test
```

# Contribute
We are always eager to improve and adopt new ideas, please feel free to make suggestions by raising a pull request.

1. Fork the Project
2. Create your Feature Branch (git checkout -b feature/AmazingFeature)
3. Commit your Changes (git commit -m 'Add some AmazingFeature')
4. Push to the Branch (git push origin feature/AmazingFeature)
5. Open a Pull Request

# Other Non Custodial Staking Implementations 

## LockRegistry

- **Author**  
  [OwlOfMoistness](https://twitter.com/OwlOfMoistness)
- **Description**  
  Provides a really cool abstraction for non-custodial staking with multiple staking instances. It also paves the way for Owl's on-chain "2FA" concept, which is potentially groundbreaking for improving user security.
- **Source**  
  https://github.com/OwlOfMoistness/erc721-lock-registry

## SmolSchool

- **Author**  
  [SmolVerse](https://www.smolverse.lol/)
- **Description**  
  AFAIK, one of the first implementations of a "non-custodial staking" that we've found. This pattern continues to be adopted across the Treasure Ecosystem, and can be found in Bridgeworld's Legions as well.
- **Source**  
  https://arbiscan.io/address/0x6325439389e0797ab35752b4f43a14c004f22a9c#code


CHANGELOG.md
## [1.0.4](https://github.com/samurisenft/erc721nes-contracts/compare/1.0.0...1.0.4) (2022-03-28)


### Bug Fixes

* **build:** check out master, clean up deploy tag ([552317e](https://github.com/samurisenft/erc721nes-contracts/commit/552317e6f3d3cadf8f53e376e66a10e40df9061c))
* **build:** do not rely on $default-branch var ([#6](https://github.com/samurisenft/erc721nes-contracts/issues/6)) ([7f72f5f](https://github.com/samurisenft/erc721nes-contracts/commit/7f72f5f84d6519e30e05cc7a11ca502372054ec8))
* **build:** fix publish dependency, fix package.json version ([#8](https://github.com/samurisenft/erc721nes-contracts/issues/8)) ([2c44e31](https://github.com/samurisenft/erc721nes-contracts/commit/2c44e31c118f0431a8e2d718ae49b889634559f7))
* **build:** run npm version patch instead of just version ([#9](https://github.com/samurisenft/erc721nes-contracts/issues/9)) ([40b080d](https://github.com/samurisenft/erc721nes-contracts/commit/40b080d166f92a65fd1843d9902e815c0ee9679e))
* **build:** set git credentials before npm version ([e27f815](https://github.com/samurisenft/erc721nes-contracts/commit/e27f815abbd0cc44eb28b26f17d44cc27545633f))
* **build:** use force push from ci ([4b39783](https://github.com/samurisenft/erc721nes-contracts/commit/4b39783bd418f9c81985e4ea38447c9b77a837a7))


### Features

* **build:** add Github workflow for running tests ([#5](https://github.com/samurisenft/erc721nes-contracts/issues/5)) ([af2c821](https://github.com/samurisenft/erc721nes-contracts/commit/af2c821335045c61125b32fab3b31f6d7082b1d7))
* **build:** add publishing workflow ([#7](https://github.com/samurisenft/erc721nes-contracts/issues/7)) ([f042102](https://github.com/samurisenft/erc721nes-contracts/commit/f04210260ac1aa965e8bc205c0172ab5102f6c67))
* **build:** support minor/major publish ([dba0612](https://github.com/samurisenft/erc721nes-contracts/commit/dba0612642927392bdde6fab77e43c4eb97d53cf))



# 1.0.0 (2022-03-27)





test/test.js
// We import Chai to use its asserting functions here.
const Web3 = require("web3");
const { soliditySha3 } = require("web3-utils");
const { privateKeyServerSide, infuraProjectId, mnemonic, etherscanApiKey } = require('../secrets.json');
const { use, expect } = require("chai");
const { ethers } = require("hardhat");

async function mineNBlocks(n) {
  for (let index = 0; index < n; index++) {
    await ethers.provider.send('evm_mine');
  }
}

describe("Token contract", function () {
  let Token;
  let test721Token;
  let testStakingController;
  let owner;
  let addr1;
  let addr2;
  let addrs;
  
  beforeEach(async function () {
    TokenFactory = await ethers.getContractFactory("ERC721NESTestImpl");
    [owner, addr1, addr2, ...addrs] = await ethers.getSigners();

    test721Token = await TokenFactory.deploy();

    TokenFactory = await ethers.getContractFactory("TestStakingController");
    [owner, addr1, addr2, ...addrs] = await ethers.getSigners();

    TokenFactory = await ethers.getContractFactory("TestStakingController");
    testStakingController = await TokenFactory.deploy(test721Token.address, 1);    
    await test721Token.setStakingController(testStakingController.address);

  });

    it("calling mint and stake should mint happy case", async function () {
      const trx = await test721Token.connect(addr1).mint(1);        
      const addr1Balance = await test721Token.balanceOf(addr1.address);
      expect(await addr1Balance.toNumber()).to.equal(1);
      expect(await test721Token.isStaked(0)).to.equal(false);
    });

    it("staking acrrues balance at correct rate when minting and then staking", async function () {
      await test721Token.connect(addr1).mint(1);
      const addr1Balance = await test721Token.balanceOf(addr1.address);
      expect(await addr1Balance.toNumber()).to.equal(1);
      expect(await testStakingController.getStakingRewards(0)).to.equal(0);
      await testStakingController.connect(addr1).stake(0);
      await mineNBlocks(10);
      expect(await testStakingController.getStakingRewards(0)).to.equal(10);
      await mineNBlocks(10);
      expect(await testStakingController.getStakingRewards(0)).to.equal(20);
    });

    it("staking acrrues balance at correct rate when performing mintAndStake", async function () {
      console.log(`owner ${owner.address}`);
      await test721Token.connect(addr1).mintAndStake(2);
      const addr1Balance = await test721Token.balanceOf(addr1.address);
      expect(await addr1Balance.toNumber()).to.equal(2);
      // expect(await testStakingController.getStakingRewards(0)).to.equal(0);
      // expect(await testStakingController.getStakingRewards(1)).to.equal(0);
      // await mineNBlocks(10);
      // expect(await testStakingController.getStakingRewards(0)).to.equal(10);
      // expect(await testStakingController.getStakingRewards(1)).to.equal(20);
      // await mineNBlocks(10);
      // expect(await testStakingController.getStakingRewards(0)).to.equal(20);
      // expect(await testStakingController.getStakingRewards(1)).to.equal(20);
    });

    it("staking / unstaking from staking controller contract works", async function () {
      const trx = await test721Token.connect(addr1).mint(1);
      const addr1Balance = await test721Token.balanceOf(addr1.address);
      expect(await addr1Balance.toNumber()).to.equal(1);

      const trx2 = await testStakingController.connect(addr1).stake(0);
      await expect(test721Token.transferFrom(addr1.address, addr2.address, 0)).to.be.revertedWith("You can not transfer a staked token");
      const trx3 = await testStakingController.connect(addr1).unstake(0);
    });
});  


test/testERC721NESImpl.js
// We import Chai to use its asserting functions here.
const Web3 = require("web3");
const { use, expect } = require("chai");
const { ethers } = require("hardhat");

async function mineNBlocks(n) {
  for (let index = 0; index < n; index++) {
    await ethers.provider.send('evm_mine');
  }
}

describe("Token contract", function () {
  let Token;
  let test721Token;
  let testStakingController;
  let owner;
  let addr1;
  let addr2;
  let addrs;
  
  beforeEach(async function () {
    [owner, addr1, addr2, ...addrs] = await ethers.getSigners();
    TokenFactory = await ethers.getContractFactory("ERC721NESImpl");
    test721Token = await TokenFactory.deploy();

    TokenFactory = await ethers.getContractFactory("ERC721NESStakingController");

    testStakingController = await TokenFactory.deploy(test721Token.address, 1);    
    await test721Token.setStakingController(testStakingController.address);
  });

    it("Staking Contoller Impl - calling mint and stake should mint happy case", async function () {
      const trx = await test721Token.connect(addr1).mint(1);        
      const addr1Balance = await test721Token.balanceOf(addr1.address);
      expect(await addr1Balance.toNumber()).to.equal(1);
      expect(await test721Token.isStaked(0)).to.equal(false);
    });

    it("Staking Contoller Impl - staking acrrues balance at correct rate when minting and then staking", async function () {
      await test721Token.connect(addr1).mint(1);
      const addr1Balance = await test721Token.balanceOf(addr1.address);
     
      expect(await addr1Balance.toNumber()).to.equal(1);
      expect(await testStakingController.getStakingRewards(0)).to.equal(0);
     
      await testStakingController.connect(addr1).stake(0);
      await mineNBlocks(10);
     
      expect(await testStakingController.getStakingRewards(0)).to.equal(10);
     
      await mineNBlocks(10);
     
      expect(await testStakingController.getStakingRewards(0)).to.equal(20);
     
      await mineNBlocks(10);
     
      expect(await testStakingController.getStakingRewards(0)).to.equal(30);
    });

    it("Staking Contoller Impl - staking prevents transfer, unstakeing allows transfer", async function () {
      const trx = await test721Token.connect(addr1).mint(1);
      let addr1Balance = await test721Token.balanceOf(addr1.address);
      let addr2Balance = await test721Token.balanceOf(addr2.address);

      expect(await addr1Balance.toNumber()).to.equal(1);
      expect(await addr2Balance.toNumber()).to.equal(0);

      const trx2 = await testStakingController.connect(addr1).stake(0);

      await expect(test721Token.transferFrom(addr1.address, addr2.address, 0)).to.be.revertedWith("You can not transfer a staked token");

      const trx3 = await testStakingController.connect(addr1).unstake(0);
      await test721Token.connect(addr1).transferFrom(addr1.address, addr2.address, 0);
      addr1Balance = await test721Token.balanceOf(addr1.address);
      addr2Balance = await test721Token.balanceOf(addr2.address);
      
      expect(await addr1Balance.toNumber()).to.equal(0);
      expect(await addr2Balance.toNumber()).to.equal(1);

    });


});  

test/testERC721NESTestImpl.js
// We import Chai to use its asserting functions here.
const Web3 = require("web3");
const { soliditySha3 } = require("web3-utils");
const { privateKeyServerSide, infuraProjectId, mnemonic, etherscanApiKey } = require('../secrets.json');
const { use, expect } = require("chai");
const { ethers } = require("hardhat");

async function mineNBlocks(n) {
  for (let index = 0; index < n; index++) {
    await ethers.provider.send('evm_mine');
  }
}

describe("Token contract", function () {
  let Token;
  let test721Token;
  let testStakingController;
  let owner;
  let addr1;
  let addr2;
  let addrs;
  
  beforeEach(async function () {
    [owner, addr1, addr2, ...addrs] = await ethers.getSigners();
    TokenFactory = await ethers.getContractFactory("ERC721NESTestImpl");
    test721Token = await TokenFactory.deploy();

    TokenFactory = await ethers.getContractFactory("ERC721NESTestStakingController");

    testStakingController = await TokenFactory.deploy(test721Token.address, 1);    
    await test721Token.setStakingController(testStakingController.address);
  });

    it("Staking Contoller Impl - calling mint and stake should mint happy case", async function () {
      const trx = await test721Token.connect(addr1).mint(1);        
      const addr1Balance = await test721Token.balanceOf(addr1.address);
      expect(await addr1Balance.toNumber()).to.equal(1);
      expect(await test721Token.isStaked(0)).to.equal(false);
    });

    it("Staking Contoller Impl - staking acrrues balance at correct rate when minting and then staking", async function () {
      await test721Token.connect(addr1).mint(1);
      const addr1Balance = await test721Token.balanceOf(addr1.address);
     
      expect(await addr1Balance.toNumber()).to.equal(1);
      expect(await testStakingController.getStakingRewards(0)).to.equal(0);
     
      await testStakingController.connect(addr1).stake(0);
      await mineNBlocks(10);
     
      expect(await testStakingController.getStakingRewards(0)).to.equal(10);
     
      await mineNBlocks(10);
     
      expect(await testStakingController.getStakingRewards(0)).to.equal(20);
     
      await mineNBlocks(10);
     
      expect(await testStakingController.getStakingRewards(0)).to.equal(30);
    });

    it("Staking Contoller Impl - staking prevents transfer, unstakeing allows transfer", async function () {
      const trx = await test721Token.connect(addr1).mint(1);
      let addr1Balance = await test721Token.balanceOf(addr1.address);
      let addr2Balance = await test721Token.balanceOf(addr2.address);

      expect(await addr1Balance.toNumber()).to.equal(1);
      expect(await addr2Balance.toNumber()).to.equal(0);

      const trx2 = await testStakingController.connect(addr1).stake(0);

      await expect(test721Token.transferFrom(addr1.address, addr2.address, 0)).to.be.revertedWith("You can not transfer a staked token");

      const trx3 = await testStakingController.connect(addr1).unstake(0);
      await test721Token.connect(addr1).transferFrom(addr1.address, addr2.address, 0);
      addr1Balance = await test721Token.balanceOf(addr1.address);
      addr2Balance = await test721Token.balanceOf(addr2.address);
      
      expect(await addr1Balance.toNumber()).to.equal(0);
      expect(await addr2Balance.toNumber()).to.equal(1);

    });


});  


scripts/deploy.js
const {ContractFactory, utils} = require('ethers');
const {MockProvider} = require('@ethereum-waffle/provider');
const {waffleChai} = require('@ethereum-waffle/chai');
const {deployMockContract} = require('@ethereum-waffle/mock-contract');
const { use } = require('chai');

use(waffleChai)

async function main() {
    const [deployer] = await ethers.getSigners();
  
    console.log("Deploying contracts with the account:", deployer.address);
    console.log("Account balance:", (await deployer.getBalance()).toString());
    const ERC721NESTestImpl = await ethers.getContractFactory("ERC721NESTestImpl");
    const erc721NESTestImpl = await ERC721NESTestImpl.deploy();
    console.log("Token address:", erc721NESTestImpl.address);
    const contract = await eRC721NESTestImpl.attach(eRC721NESTestImpl.address);
    
    const trx2 = await contract.setSignatureVerifier("0x2f87a2ef76a3389f3b95da865be8667fe53d956a");
    console.log(trx2);

  }
  
  main()
    .then(() => process.exit(0))
    .catch((error) => {
      console.error(error);
      process.exit(1);
    });

scripts/getBalance.js
const Web3 = require("web3");
const { soliditySha3 } = require("web3-utils");
const { privateKeyServerSide, infuraProjectId, mnemonic, etherscanApiKey } = require('../secrets.json');
const { use, expect } = require("chai");
const { ethers } = require("hardhat");

async function main() {
  const [deployer] = await ethers.getSigners();
  var web3 = new Web3('https://rinkebys.infura.io/v3/908e29044bcc435e9183aae2ba70e625'); // your geth
  console.log("Deploying contracts with the account:", deployer.address);
  const Test721 = await ethers.getContractFactory("Test721");
//    const contract = await Test721.attach("0xE09bB83215843EB0B246F0296833E676bAE46Ec4");
  const contract = await Test721.attach("0x220BC1980Fc2251A6Dbf66F95a39da42534450b8");
  //trx = await contract.stake(0);
  //console.log(trx);

  let balance = await contract.getCurrentAdditionalBalance(0);
  console.log(balance);


  }
  
  main()
    .then(() => process.exit(0))
    .catch((error) => {
      console.error(error);
      process.exit(1);
    });

scripts/mint.js
const Web3 = require("web3");
const { soliditySha3 } = require("web3-utils");
const { privateKeyServerSide, infuraProjectId, mnemonic, etherscanApiKey } = require('../secrets.json');
const { use, expect } = require("chai");
const { ethers } = require("hardhat");

async function main() {
    const [deployer] = await ethers.getSigners();
    var web3 = new Web3('https://rinkebys.infura.io/v3/908e29044bcc435e9183aae2ba70e625'); // your geth
    console.log("Deploying contracts with the account:", deployer.address);
    const Test721 = await ethers.getContractFactory("Test721");
    const contract = await Test721.attach("0xE09bB83215843EB0B246F0296833E676bAE46Ec4");
  
    
    //let trx = await contract.unpause();
    //console.log(trx);

    const sig = '0x2f87a2ef76a3389f3b95da865be8667fe53d956a';

    trx = await contract.setSignatureVerifier(sig);
    console.log(trx);

    
    const nonce = 0;
    const localHash = soliditySha3(deployer['address'] , nonce);
    
    console.log(`localHash: ${localHash}`);
    const signature = web3.eth.accounts.sign(localHash, privateKeyServerSide)['signature'];
    
    console.log(`signature ${signature}`)
    trx = await contract.allowListMint(signature, nonce, {value: '47000000000000000' });
    console.log(trx);
    const receipt = await trx.wait();

    console.log(receipt);

    // const nonce = 0;
    // const localHash = soliditySha3(deployer['address'] , nonce);
    
    // console.log(`localHash: ${localHash}`);
    // const signature = web3.eth.accounts.sign(localHash, privateKeyServerSide)['signature'];

    // console.log(`signature ${signature}`)
    // const trx4 = await contract.allowListMintAndStake(signature, nonce, {value: '47000000000000000' });
    // console.log(trx4);

    
    trx = await contract.stake(0);
    console.log(trx);

    let balance = await contract.getStakedCumulativeBalance(0);
    console.log(balance);


  }
  
  main()
    .then(() => process.exit(0))
    .catch((error) => {
      console.error(error);
      process.exit(1);
    });